Assets/Mirror/Components/NetworkProximityChecker.cs

```diff
@@ -104,55 +71,25 @@ public override void OnRebuildObservers(HashSet<NetworkConnection> observers, bo
-            if (hitCount == hitsBuffer3D.Length) Debug.LogWarning("NetworkProximityChecker's OverlapSphere test for " + name + " has filled the whole buffer(" + hitsBuffer3D.Length + "). Some results might have been omitted. Consider increasing buffer size.");
-
-            for (int i = 0; i < hitCount; i++)
-            {
-                Collider hit = hitsBuffer3D[i];
-                // collider might be on pelvis, often the NetworkIdentity is in a parent
-                // (looks in the object itself and then parents)
-                NetworkIdentity identity = hit.GetComponentInParent<NetworkIdentity>();
-                // (if an object has a connectionToClient, it is a player)
-                if (identity != null && identity.connectionToClient != null)
-                {
-                    observers.Add(identity.connectionToClient);
-                }
-            }
-        }
-
-        void Add2DHits(HashSet<NetworkConnection> observers)
-        {
-            // cast without allocating GC for maximum performance
-            int hitCount = Physics2D.OverlapCircleNonAlloc(transform.position, visRange, hitsBuffer2D, castLayers);
-            if (hitCount == hitsBuffer2D.Length) Debug.LogWarning("NetworkProximityChecker's OverlapCircle test for " + name + " has filled the whole buffer(" + hitsBuffer2D.Length + "). Some results might have been omitted. Consider increasing buffer size.");
-
-            for (int i = 0; i < hitCount; i++)
+            // 'transform.' calls GetComponent, only do it once
+            Vector3 position = transform.position;
+
+            // brute force distance check
+            // -> only player connections can be observers, so it's enough if we
+            //    go through all connections instead of all spawned identities.
+            // -> compared to UNET's sphere cast checking, this one is orders of
+            //    magnitude faster. if we have 10k monsters and run a sphere
+            //    cast 10k times, we will see a noticeable lag even with physics
+            //    layers. but checking to every connection is fast.
+            foreach (NetworkConnectionToClient conn in NetworkServer.connections.Values)
             {
-                Collider2D hit = hitsBuffer2D[i];
-                // collider might be on pelvis, often the NetworkIdentity is in a parent
-                // (looks in the object itself and then parents)
-                NetworkIdentity identity = hit.GetComponentInParent<NetworkIdentity>();
-                // (if an object has a connectionToClient, it is a player)
-                if (identity != null && identity.connectionToClient != null)
+                if (conn != null && conn.identity != null)
                 {
-                    observers.Add(identity.connectionToClient);
+                    // check distance
+                    if (Vector3.Distance(conn.identity.transform.position, position) < visRange)
```

Comment:
```suggestion
                    // SqrMagnitude is faster than Distance per Unity docs
                    // https://docs.unity3d.com/ScriptReference/Vector3-sqrMagnitude.html
                    if (Vector3.SqrMagnitude(conn.identity.transform.position, position) < visRange * visRange)
```

Reply:
Let's do this one if it shows up in the profiler.
KISS first :)

Reply:
Sure...just actually watch for it :)

Reply:
To be fair, it amounts to very little on a server or stout PC, and since Prox is a server-side process it won't really matter other than to those who obsess over such things.

Where it shows up is on clients, especially mobile devices.  This why I've used SqrMagnitude in the experimental NetworkTransform since that does do work on clients.

Coburn did some profiling:
https://github.com/vis2k/Mirror/pull/1514#issuecomment-591319653

Reply:
We can write performance tests for this later to see if it is worth changing.

Reply:
> Where it shows up is on clients, especially mobile devices. This is why I've used SqrMagnitude in the experimental NetworkTransform since that does do work on clients.
> 
> Coburn did some profiling:
> [#1514 (comment)](https://github.com/vis2k/Mirror/pull/1514#issuecomment-591319653)
Just gonna say if you don't care about mobile devices, then sure, that's fine. Just keep in mind a little micro-optimization for mobile devices and by extension, the Nintendo Switch, might do wonders since some instructions on ARM aren't hardware accelerated.

Reply:
I think this check is only done on the server, so we don't have to worry about mobile devices here.

Reply:
@James-Frowen while it is a server-only function, you have a oversight: what if the mobile device is the host? Usually you shouldn't use mobile devices as hosts but our userbase seems to ignore this fact. So if you're using a mobile device as a client, then sure, it won't affect it. But if you're a host client on the server, then you will get the benefit.